### ISSUE TITLE
Question: Extended Cron Syntax Support

### ISSUE THREAD

**Issue by machenity** [kelektiv/node-cron#435]
Does this package have any plan to support non-standard cron syntax?
Like, https://en.wikipedia.org/wiki/Cron#Non-standard_characters

I am using this package really well but I needed to use extended syntax for some reasons.
NowI extended the codes myself but it would be great if these syntaxes are supported natively.

**Comment by ncb000gt** [comment]
Thanks for the resource. I've gone back and forth on supporting extended characters before. Mostly my issue is that I just don't have time to build out the support myself. If you've done it I'd review a pr for sure. Thanks!

**Comment by Deepak13245** [comment]
Hi,
I would like to take this up. Please assign it to me

**Comment by ncb000gt** [comment]
@Deepak13245 did you ever end up making any progress on this task?

**Comment by ramvinoth** [comment]
Any update on this? Thanks in advance.

**Comment by sheerlox** [comment]
I propose we implement the following extended characters list:

> `?`: The question mark is substituted with the time of initialization. For example, `? ? * * * *` would be substituted with `25 8 * * * *` if the time is 08:25 at the time of `new Cron('? ? * * * *', <...>)`. The question mark can be used in any field.

> `L`: The letter `L` can be used in the day of the month field to indicate the last day of the month. When used in the day of the week field in conjunction with the `#` character, it denotes the last specific weekday of the month. For example, `5#L` represents the last Friday of the month.

> `#`: The `#` character specifies the “nth” occurrence of a particular day within a month. For example, supplying `5#2` in the day-of-week field signifies the second Friday of the month. This can be combined with ranges and supports day names. For instance, `MON-FRI#2` would match the Monday through Friday of the second week of the month.

*source: [`croner` documentation](https://croner.56k.guru/usage/pattern.html)*

this only leaves out the `W` character for day-of-month, which "is used to specify the weekday (Monday-Friday) nearest the given day", and it doesn't seem like a priority to implement. this could be nice to have, but we can probably leave it out for now.

**Comment by mfcarroll** [comment]
> > `#`: The `#` character specifies the “nth” occurrence of a particular day within a month. For example, supplying `5#2` in the day-of-week field signifies the second Friday of the month. This can be combined with ranges and supports day names. For instance, `MON-FRI#2` would match the Monday through Friday of the second week of the month.

I'd find this particular extended syntax incredibly useful for my project. Just wondering if there's been any progress here, or if there's anything I can do to help make it happen. Thanks!

**Comment by sheerlox** [comment]
> I'd find this particular extended syntax incredibly useful for my project. Just wondering if there's been any progress here, or if there's anything I can do to help make it happen. Thanks!

Hey there! Because we maintainers have had a very low bandwidth lately, we are waiting for volunteers to pick up on some of the waiting feature requests. If you'd be interested in taking a shot at implementing the extended cron syntax, I'd be glad to entertain any PR and give a hand if necessary!

**Comment by alexanderkrum** [comment]
Picking this up

**Comment by alexanderkrum** [comment]
Yeah will do tmr. I am thinking of separating special chars handling to a separate class. We can discuss it in the draft.

**Comment by sheerlox** [comment]
Comment from @bao-multiIT on https://github.com/kelektiv/node-cron/pull/877#issuecomment-2609295510:
> i need the L syntax to run on the n-th last day of each month, for example L-1 would run 1 day before the last day of the month, if the month has 30 days then it would run on 29th, if the month has 31 days then it would run on 30th
> 
> seems like this syntax already exists, i can see it on [this cron expression generator](https://www.freeformatter.com/cron-expression-generator-quartz.html), but it is not implemented in this library yet
> 
> does anyone have an idea for a workaround while waiting for this feature?

**Comment by sheerlox** [comment]
Since our current library version doesn't (yet) support the `L` syntax for specifying expressions like "Nth last day of the month," a practical workaround is to schedule the job to run on the days when the Nth last day can occur and then, within your `onTick` function, check if the current day is indeed the Nth last day of the month. Something along those lines should do the trick:

```js
import { CronJob } from 'cron';

const Nth = 1; // For example, Nth = 1 for the second to last day

const job = CronJob.from({
  '0 0 0 28-31 * *', // At 00:00:00 on days 28-31 every month
  onTick: function () {
    const today = new Date();
    const lastDayOfMonth = new Date(today.getFullYear(), today.getMonth() + 1, 0).getDate();
    const daysUntilEndOfMonth = lastDayOfMonth - today.getDate();

    if (daysUntilEndOfMonth === Nth) {
      // Today is the Nth last day of the month
      // Place your task execution code here
      console.log(`Executing task on N=${Nth} last day of the month: ${today.toDateString()}`);
    }
  },
  start: true
});
```

**Comment by bao-multiIT** [comment]
> Since this library doesn't (yet) support the `L` syntax for specifying expressions like "Nth last day of the month," a practical workaround is to schedule the job to run on the days when the Nth last day can occur and then, within your `onTick` function, check if the current day is indeed the Nth last day of the month. Something along those lines should do the trick:
> 
> import { CronJob } from 'cron';
> 
> const Nth = 1; // For example, Nth = 1 for the second to last day
> 
> const job = CronJob.from({
>   '0 0 0 28-31 * *', // At 00:00:00 on days 28-31 every month
>   onTick: function () {
>     const today = new Date();
>     const lastDayOfMonth = new Date(today.getFullYear(), today.getMonth() + 1, 0).getDate();
>     const daysUntilEndOfMonth = lastDayOfMonth - today.getDate();
> 
>     if (daysUntilEndOfMonth === Nth) {
>       // Today is the Nth last day of the month
>       // Place your task execution code here
>       console.log(`Executing task on N=${Nth} last day of the month: ${today.toDateString()}`);
>     }
>   },
>   start: true
> });

makes sense, thank you for helping!

**Comment by hd-esteban** [comment]
@sheerlox do you've any update on this ticket? this would be really helpful
